### PR TITLE
Delete participant and host feature

### PR DESCRIPTION
**Pull request by Kajol-Kumari**
Feat:#2648 
Fixed: #2664 

Features added:
1. Participant team can delete it's member.
2. Host team can delete it's member

GIF for Participant team:
ezgif com-video-to-gif (3)

GIF for Host team:
ezgif com-video-to-gif (4)

Fix of #2664 
GIF for showing trash icon feature:
ezgif com-video-to-gif (8)

### CODE REVIEW

File: apps/hosts/views.py [hunk 1]
```diff
@@ -254,6 +254,47 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+                response_data, status=status.HTTP_406_NOT_ACCEPTABLE
+            )
+        else:
+            challenge_host.delete()
+            return Response(status=status.HTTP_204_NO_CONTENT)
+    else:
+        response_data = {
+            "error": "Sorry, you do not have permissions to remove this challenge host"
+        }
+        return Response(response_data, status=status.HTTP_401_UNAUTHORIZED)
```
Comment (nikochiko) [inline apps/hosts/views.py]: Similar functionality is implemented here: https://github.com/Cloud-CV/EvalAI/blob/master/apps/hosts/views.py#L204-L206 
It also has URL assigned to it. Could you implement this there itself?
The linked one doesn't check the permissions of the user before deleting. I might be missing something, but as I understand the code, that should be a bug. Implementing these checks there would be helpful.

Please let me know if I am missing something.

File: tests/unit/hosts/test_views.py [hunk 2]
```diff
@@ -573,3 +573,135 @@ def test_when_requesting_user_is_not_part_of_team(self):
         response = self.client.post(self.url, self.data)
         self.assertEqual(response.data, expected)
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+
+class DeleteChallengeHostFromTeamTest(BaseAPITestClass):
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: How is this different from `DeleteParticularChallengeHost`? (https://github.com/Cloud-CV/EvalAI/blob/master/tests/unit/hosts/test_views.py#L429)

File: tests/unit/hosts/test_views.py [hunk 3]
```diff
@@ -573,3 +573,135 @@ def test_when_requesting_user_is_not_part_of_team(self):
...
+        )
+
+        self.challenge_host3 = ChallengeHost.objects.create(
+            user=self.user3,
+            team_name=self.challenge_host_team,
+            status=ChallengeHost.ACCEPTED,
+            permissions=ChallengeHost.ACCEPTED,
+        )
+
+        self.client.force_authenticate(user=self.user3)
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: Can we define all this in the `setUp`?

File: tests/unit/hosts/test_views.py [hunk 4]
```diff
@@ -573,3 +573,135 @@ def test_when_requesting_user_is_not_part_of_team(self):
...
+
+        self.url = reverse_lazy(
+            "hosts:delete_challenge_host_from_team",
+            kwargs={
+                "challenge_host_team_pk": self.challenge_host_team.pk,
+                "challenge_host_pk": self.invite_user.pk,
+            },
+        )
+
+    def test_challenge_host_does_not_exist_in_team(self):
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: How about renaming this to `test_delete_challenge_host_when_host_does_not_exist` to follow convention?

File: tests/unit/hosts/test_views.py [hunk 5]
```diff
@@ -573,3 +573,135 @@ def test_when_requesting_user_is_not_part_of_team(self):
...
+            },
+        )
+
+        expected = {"error": "ChallengeHost does not exist"}
+
+        response = self.client.delete(self.url, {})
+        self.assertEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
+
+    def test_when_challenge_host_team_does_not_exist(self):
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: Similarly, here. We are naming tests with format `test_perform_functionality_when_some_condition`. We could follow the same convention.

File: tests/unit/hosts/test_views.py [hunk 6]
```diff
@@ -573,3 +573,135 @@ def test_when_requesting_user_is_not_part_of_team(self):
...
+            },
+        )
+
+        expected = {"error": "ChallengeHostTeam does not exist"}
+
+        response = self.client.delete(self.url, {})
+        self.assertEqual(response.data, expected)
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
+
+    def test_when_challenge_host_is_admin_and_wants_to_delete_himself(self):
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: Similarly for other tests.

File: frontend/tests/controllers-test/challengeHostTeamsCtrl.test.js [hunk 7]
```diff
@@ -415,7 +415,7 @@ describe('Unit tests for challenge host team controller', function () {
         });
 
          it('open dialog to confirm delete', function () {
-            var participantTeamId = 1;
+            var challengeHostTeamId = 1;
```
[the remark this answers is not in the document]
Comment (Kajol-Kumari) [reply frontend/tests/controllers-test/challengeHostTeamsCtrl.test.js]: > This was a typo correction, right?
> If yes, then I think `hostTeamId` would be more consistent.

yes @nikochiko it was a typo correction. I will change the names to more appropriate name. Thanks for your suggestion.

File: tests/unit/hosts/test_views.py (hunk 2 above)
Comment (Kajol-Kumari) [reply tests/unit/hosts/test_views.py]: > How is this different from `DeleteParticularChallengeHost`? (https://github.com/Cloud-CV/EvalAI/blob/master/tests/unit/hosts/test_views.py#L429)

@nikochiko in DeleteParticularChallengeHost, there was no check if admin decided to delete himself and so it will delete the whole host team if admin tried to delete himself. So i wrote another test which checks if it's the admin who tries to delete himself/herself and if so then show him/her an error message and deny the request by showing appropriated message. 

Also i was not sure why DeleteParticularChallengeHost was written so instead of editing it i preferred writing a new one.

File: apps/hosts/views.py (hunk 1 above)
Comment (Kajol-Kumari) [reply apps/hosts/views.py]: > Similar functionality is implemented here: https://github.com/Cloud-CV/EvalAI/blob/master/apps/hosts/views.py#L204-L206
> It also has URL assigned to it. Could you implement this there itself?
> The linked one doesn't check the permissions of the user before deleting. I might be missing something, but as I understand the code, that should be a bug. Implementing these checks there would be helpful.
> 
> Please let me know if I am missing something.

@nikochiko That should have been the code for deleting host member but when i used the url associated with that particular function, it let me to delete the whole team as it is not performing any kind of check. If that is not written for any other purpose i'll edit that only to get all the checks done and then perform the deletion. Please let me know if can do it.

File: frontend/src/views/web/challenge-host-teams.html [hunk 8]
```diff
@@ -33,7 +33,9 @@
                             </div>
                             <div class="collapsible-body">
                                 <h6><b>Team Members:</b></h6>
-                                <p ng-repeat="member in value.members">{{member.user}}</p>
+                                <p ng-repeat="member in value.members">{{member.user}}
+                                    <a class="show-member-title pointer" ng-click="challengeHostTeams.confirmMemberDelete($event, value.id, member.id)"><i class="fa fa-trash list-icon text-highlight w-300 right"></i></a>
+                                </p>
```
[the remark this answers is not in the document]
Comment (Kajol-Kumari) [reply frontend/src/views/web/challenge-host-teams.html]: > As only the creators of the team can remove members, what do you think about only displaying the trash icon to team creators?

@nikochiko Thanks for pointing this out. It had skipped out of my mind.

File: tests/unit/hosts/test_views.py (hunk 3 above)
Comment (Kajol-Kumari) [reply tests/unit/hosts/test_views.py]: > Can we define all this in the `setUp`?

@nikochiko in one test class only one setUp is supported and if i put all these in the main setUp the particular function is unable to access it and throwing error.

File: frontend/src/views/web/challenge-host-teams.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kajol-Kumari) [reply frontend/src/views/web/challenge-host-teams.html]: > As only the creators of the team can remove members, what do you think about only displaying the trash icon to team creators?

@nikochiko Added that feature.Only the creator of the team will see that trash icon.

File: apps/hosts/urls.py [hunk 9]
```diff
@@ -33,6 +33,11 @@
         views.remove_self_from_challenge_host_team,
         name="remove_self_from_challenge_host_team",
     ),
+    url(
+        r"^challenge_host_team/(?P<challenge_host_team_pk>[0-9]+)/challenge_host/(?P<challenge_host_pk>[0-9]+)/delete$",
+        views.delete_challenge_host_from_team,
+        name="delete_challenge_host_from_team",
+    ),
```
Comment (nikochiko) [inline apps/hosts/urls.py]: This would not be required as there is already this URL: https://github.com/Cloud-CV/EvalAI/blob/master/apps/hosts/urls.py#L26-L30

File: apps/hosts/views.py (hunk 1 above)
Comment (nikochiko) [reply apps/hosts/views.py]: It is probably a bug in that case.
Go ahead and add the changes there itself instead of having two URLs/views for the same task.
For keeping record, you can create an issue stating the bug and attaching a GIF.
cc: @RishabhJain2018
reference: https://github.com/Cloud-CV/EvalAI/blob/master/apps/hosts/views.py#L204-L206

File: frontend/src/js/controllers/challengeHostTeamsCtrl.js [hunk 10]
```diff
@@ -385,6 +385,87 @@
             }, function() {});
         };
 
+// Delete a particular host
+vm.confirmMemberDelete = function(ev, hostTeamId, HostId) {
```
Comment (nikochiko) [inline frontend/src/js/controllers/challengeHostTeamsCtrl.js]: Can we change `HostId` -> `hostId` to maintain `camelCase`?

File: frontend/src/js/controllers/teamsCtrl.js [hunk 11]
```diff
@@ -324,6 +325,87 @@
             });
         };
 
+// Delete only a particular member
+vm.confirmMemberDelete = function(ev, ParticipantTeamId, MemberId) {
```
Comment (nikochiko) [inline frontend/src/js/controllers/teamsCtrl.js]: Can we use `camelCase` here?

File: frontend/src/views/web/challenge-host-teams.html [hunk 12]
```diff
@@ -33,7 +33,9 @@
                             </div>
                             <div class="collapsible-body">
                                 <h6><b>Team Members:</b></h6>
-                                <p ng-repeat="member in value.members">{{member.user}}</p>
+                                <p ng-repeat="member in value.members">{{member.user}}
+                                    <a class="show-member-title pointer" ng-if="value.created_by == web.name" ng-click="challengeHostTeams.confirmMemberDelete($event, value.id, member.id)"><i class="fa fa-trash list-icon text-highlight w-300 right"></i></a>
```
Comment (nikochiko) [inline frontend/src/views/web/challenge-host-teams.html]: Nice work with the `ng-if`. 
I am not familiar with frontend so I don't know if `web.name` is valid. 
Cloning and running it on my local is taking some time. Have you tried it on your side?

File: frontend/src/views/web/teams.html [hunk 13]
```diff
@@ -31,7 +31,9 @@
                             </div>
                             <div class="collapsible-body">
                                 <h6><b>Team Members:</b></h6>
-                                <p ng-repeat="member in value.members">{{member.member_name}}</p>
+                                <p ng-repeat="member in value.members">{{member.member_name}}  
+                                    <a class="show-member-title pointer" ng-if="value.created_by == web.name" ng-click="teams.confirmMemberDelete($event, value.id, member.id)"><i class="fa fa-trash list-icon text-highlight w-300 right"></i></a>
```
Comment (nikochiko) [inline frontend/src/views/web/teams.html]: Nice work :tada:

File: frontend/tests/controllers-test/challengeHostTeamsCtrl.test.js [hunk 14]
```diff
@@ -424,11 +424,36 @@ describe('Unit tests for challenge host team controller', function () {
...
+        beforeEach(function () {
+            spyOn($mdDialog, 'show').and.callFake(function () {
+                var deferred = $injector.get('$q').defer();
+                return deferred.promise;
+            });
+        });
+
+        it('open dialog to confirm delete', function () {
+            var hostTeamId = 1;
+            var HostId = 1;
```
Comment (nikochiko) [inline frontend/tests/controllers-test/challengeHostTeamsCtrl.test.js]: Can we use `camelCase` here?

File: frontend/tests/controllers-test/teamsCtrl.test.js [hunk 15]
```diff
@@ -327,6 +327,31 @@ describe('Unit tests for teams controller', function () {
...
+        beforeEach(function () {
+            spyOn($mdDialog, 'show').and.callFake(function () {
+                var deferred = $injector.get('$q').defer();
+                return deferred.promise;
+            });
+        });
+
+        it('open dialog to confirm delete', function () {
+            var participantTeamId = 1;
+            var MemberId = 1;
```
Comment (nikochiko) [inline frontend/tests/controllers-test/teamsCtrl.test.js]: `memberId`?

File: tests/unit/hosts/test_views.py (hunk 2 above)
Comment (nikochiko) [reply tests/unit/hosts/test_views.py]: So as in [this](https://github.com/Cloud-CV/EvalAI/pull/2660/files#r377025730) conversation, it is probably a bug. 
But having two views for performing the same action would be confusing - see [Single Responsibility Principle](https://codeburst.io/the-single-responsibility-principle-2c09c279fce9?gi=3d98d7da3669). 
And same thing with two classes for testing the same functionality, what do you think about merging these two classes?

File: tests/unit/hosts/test_views.py (hunk 3 above)
Comment (nikochiko) [reply tests/unit/hosts/test_views.py]: @Kajol-Kumari
>
> @nikochiko if i put all these in the main setUp the particular function is unable to access it and throwing error.

I didn't understand what you mean. 
I am saying we could define it in [this](https://github.com/Cloud-CV/EvalAI/pull/2660/files#diff-622550d062dfae0cf9c89d9c3b1ab078R579) `setUp` method like we did for other functions. Please let me know if I am missing something.

File: tests/unit/hosts/test_views.py (hunk 2 above)
Comment (Kajol-Kumari) [reply tests/unit/hosts/test_views.py]: > So as in [this](https://github.com/Cloud-CV/EvalAI/pull/2660/files#r377025730) conversation, it is probably a bug.
> But having two views for performing the same action would be confusing - see [Single Responsibility Principle](https://codeburst.io/the-single-responsibility-principle-2c09c279fce9?gi=3d98d7da3669).
> And same thing with two classes for testing the same functionality, what do you think about merging these two classes?

I was also thinking the same but was not sure. I will just open up an issue for previously written view and url and will edit there only to make the deletion functionality work.

File: frontend/src/views/web/challenge-host-teams.html (hunk 12 above)
Comment (Kajol-Kumari) [reply frontend/src/views/web/challenge-host-teams.html]: > Nice work with the `ng-if`.
> I am not familiar with frontend so I don't know if `web.name` is valid.
> Cloning and running it on my local is taking some time. Have you tried it on your side?

yes @nikochiko . I will now update the GIF accordingly to show the changed behaviour.

File: tests/unit/hosts/test_views.py (hunk 3 above)
Comment (Kajol-Kumari) [reply tests/unit/hosts/test_views.py]: > @Kajol-Kumari
> 
> > @nikochiko if i put all these in the main setUp the particular function is unable to access it and throwing error.
> 
> I didn't understand what you mean.
> I am saying we could define it in [this](https://github.com/Cloud-CV/EvalAI/pull/2660/files#diff-622550d062dfae0cf9c89d9c3b1ab078R579) `setUp` method like we did for other functions. Please let me know if I am missing something.

I was doing the wrong way. Got my fault and fixed it. Also there were two test classes written pointing to the same url as that of "DeleteParticularChallengeHost". I have commented them along with providing a comment on the top.

File: apps/hosts/views.py [hunk 16]
```diff
@@ -202,8 +202,25 @@ def challenge_host_detail(request, challenge_host_team_pk, pk):
             )
 
     elif request.method == "DELETE":
-        challenge_host.delete()
-        return Response(status=status.HTTP_204_NO_CONTENT)
+        if challenge_host_team.created_by == request.user:
+
+            if (
+                challenge_host.user == request.user
+            ):  # when the user tries to remove himself
```
Comment (nikochiko) [inline apps/hosts/views.py]: How about 
```python
if challenge_host.user == request.user:
    ... do something
    ...
```
One-liners are easier to read.

File: apps/hosts/views.py [hunk 17]
```diff
@@ -254,6 +271,47 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+#         else:
+#             challenge_host.delete()
+#             return Response(status=status.HTTP_204_NO_CONTENT)
+#     else:
+#         response_data = {
+#             "error": "Sorry, you do not have permissions to remove this challenge host"
+#         }
+#         return Response(response_data, status=status.HTTP_401_UNAUTHORIZED)
+
+
```
Comment (nikochiko) [inline apps/hosts/views.py]: We can remove this 😄

File: tests/unit/hosts/test_urls.py [hunk 18]
```diff
@@ -14,6 +14,12 @@ def setUp(self):
             password="secret_password",
         )
 
+        self.invite_host_user = User.objects.create(
+            username="otheruser",
+            email="other@platform.com",
+            password="other_secret_password",
+        )
+
```
Comment (nikochiko) [inline tests/unit/hosts/test_urls.py]: I think we can use the already existing `self.challenge_host` instead.

File: tests/unit/hosts/test_urls.py [hunk 19]
```diff
@@ -57,18 +63,18 @@ def test_host_urls(self):
         )
 
         url = reverse_lazy(
-            "hosts:get_challenge_host_details",
+            "hosts:get_challenge_host_delete",
             kwargs={
                 "challenge_host_team_pk": self.challenge_host_team.pk,
-                "pk": self.challenge_host.pk,
+                "challenge_host_pk": self.invite_host_user.pk,
```
Comment (nikochiko) [inline tests/unit/hosts/test_urls.py]: Do we have to change from `self.challenge_host.pk` to `self.invite_host_user.pk`?

File: tests/unit/hosts/test_views.py [hunk 20]
```diff
@@ -328,116 +328,236 @@ def test_create_challenge_host_team_with_no_data(self):
...
+#             "permissions": ChallengeHost.WRITE,
+#         }
+#         response = self.client.put(self.url, self.data)
+#         self.assertEqual(response.data, expected)
+#         self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+#     def test_particular_challenge_host_update_with_no_data(self):
+#         self.data = {"status": ""}
+#         response = self.client.put(self.url, self.data)
+#         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: @Kajol-Kumari Please don't remove existing tests. The updates should work with original tests as well.
>
>  Commenting due to absence of url to these tests along with these two test cases are pointing to the same url name i.e get_challenge_host_details.

So the URL you renamed above is for both functionalities. It will go through the lines of code you wrote for deleting when it is used with `DELETE` method and will execute differently when the method is `GET`.
Your additions should be compatible with both. :)

File: tests/unit/hosts/test_views.py [hunk 21]
```diff
@@ -573,3 +693,127 @@ def test_when_requesting_user_is_not_part_of_team(self):
...
+#     def test_delete_challenge_host_when_a_challenge_host_is_successfully_removed_from_team(self):
+#         self.url = reverse_lazy(
+#             "hosts:delete_challenge_host_from_team",
+#             kwargs={
+#                 "challenge_host_team_pk": self.challenge_host_team.pk,
+#                 "challenge_host_pk": self.challenge_host2.pk,
+#             },
+#         )
+#         response = self.client.delete(self.url, {})
+#         self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
```
Comment (nikochiko) [inline tests/unit/hosts/test_views.py]: Please remove these comments.

File: apps/hosts/urls.py [hunk 22]
```diff
@@ -24,9 +24,9 @@
         name="get_challenge_host_list",
     ),
     url(
-        r"^challenge_host_team/(?P<challenge_host_team_pk>[0-9]+)/challenge_host/(?P<pk>[0-9]+)$",
-        views.challenge_host_detail,
-        name="get_challenge_host_details",
+        r"^challenge_host_team/(?P<challenge_host_team_pk>[0-9]+)/challenge_host/(?P<challenge_host_pk>[0-9]+)$",
+        views.challenge_host_delete,
+        name="get_challenge_host_delete",
```
Comment (nikochiko) [inline apps/hosts/urls.py]: Thanks for your thoughtfulness in changing the names. I agree the old names are somewhat unclear.
But I think the main purpose of the PR (to add feature to delete host/participant) is already almost complete. I think a different PR for renaming would be more convenient and for this PR, we can just focus on adding the feature.
Again, thanks for the thoughtfulness. :)

File: apps/hosts/views.py [hunk 23]
```diff
@@ -152,7 +152,7 @@ def challenge_host_list(request, challenge_host_team_pk):
 @throttle_classes([UserRateThrottle])
 @permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
 @authentication_classes((ExpiringTokenAuthentication,))
-def challenge_host_detail(request, challenge_host_team_pk, pk):
+def challenge_host_delete(request, challenge_host_team_pk, challenge_host_pk):
```
Comment (nikochiko) [inline apps/hosts/views.py]: Can you please move the renaming to another PR?

File: tests/unit/hosts/test_urls.py (hunk 19 above)
Comment (Kajol-Kumari) [reply tests/unit/hosts/test_urls.py]: > Do we have to change from `self.challenge_host.pk` to `self.invite_host_user.pk`?

No, it works for self.challenge_host.pk as well. Thanks for pointing this out.

File: apps/hosts/urls.py (hunk 22 above)
Comment (Kajol-Kumari) [reply apps/hosts/urls.py]: @nikochiko  All the suggested changes has been done and will open a new PR regarding renaming as soon as possible.Thanks a lot for your suggestions.I helped me a lot to improve my code structure and functionality.

Comment (nikochiko) [reply apps/hosts/urls.py]: Perfect!

**Comment by RishabhJain2018** [comment]
@Sanji515 Can you please review this once?
